src/store/zoom.ts

```diff
@@ -0,0 +1,151 @@
+import { createSlice, PayloadAction, createSelector } from '@reduxjs/toolkit';
+import { RootState } from '../store';
+
+interface ZoomState {
+	byId: {
+		[zoomId: string]: {
+			zoomId: string;
+			discordId?: string;
+			name: string;
+		};
+	};
+	online: string[];
+	temporary: string[];
+	hasSeenStart: boolean;
+	active: boolean;
+	joinUrl: string;
+	lastZoomTime: number;
+}
+
+const initialState: ZoomState = {
+	byId: {},
+	online: [],
+	temporary: [],
+	hasSeenStart: false,
+	active: false,
+	joinUrl: '',
+	lastZoomTime: -1,
+};
+
+type UserEventAction = PayloadAction<{ zoomId: string; name: string }>;
+type UserLinkAction = PayloadAction<{ zoomId: string; discordId: string }>;
+
+const { reducer, actions, name } = createSlice({
+	name: 'zoomInfo',
+	reducers: {
+		callStarted: (state) => {
+			state.active = true;
+			state.online = [];
+			state.temporary = [];
+		},
+		callEnded: (state) => {
+			state.active = false;
+			state.hasSeenStart = true;
+			state.online = [];
+			// remove temp users
+			state.temporary.forEach((id) => {
+				delete state.byId[id];
+			});
+			state.temporary = [];
+		},
+		userLeft: (state, { payload }: UserEventAction) => {
+			// always go for the most up to date name
+			if (state.byId.hasOwnProperty(payload.zoomId)) {
+				state.byId[payload.zoomId].name = payload.name;
+			}
+
+			if (!state.active) {
+				console.log('WARNING: Zoom participant left while meeting was inactive. Ignoring.');
+			} else {
+				state.online = state.online.filter((zoomId) => zoomId !== payload.zoomId);
+				if (state.temporary.indexOf(payload.zoomId) > -1) {
+					state.temporary = state.temporary.filter((zoomId) => zoomId !== payload.zoomId);
+					delete state.byId[payload.zoomId];
+				}
+			}
+		},
+		userJoined: (
+			state,
+			{
+				payload: { temporary, ...payload },
+			}: PayloadAction<{ temporary: boolean }> & UserEventAction,
+		) => {
+			state.byId[payload.zoomId] = {
+				...state.byId[payload.zoomId],
```

Comment:
nit: handle if the user doesn't already exist. Not sure if this will crash or not, but it's good to be safe. Ex: `...(state.byId[payload.zoomId] || {})`.

Reply:
spreading `undefined` is equivalent to spreading an empty object